### ISSUE TITLE
DLL initialization failure (Windows)

### ISSUE THREAD

**Issue by black-snow** [szwacz/electron-boilerplate#96]
Original issue: https://github.com/agnat/node_mdns/issues/147

I'm facing an issue with a (native) module. After some trouble I got it installed (`npm install --save mdns`) into my `app` folder. But when I try to use that module I get the following error:

![error](https://cloud.githubusercontent.com/assets/579733/11933538/bf56d29a-a7fc-11e5-973b-53d71d320bbb.PNG)

The troublesome line is pretty much just this:

```javascript
module.exports = require('..\build\Release\dns_sd_bindings');
```

There is a `dns_sd_bindings.node` file in the target directory.

I tried to `eletron-rebuild` the whole thing but that didn't help. Any guesses?

Win 10, latest boilerplate version

**Comment by szwacz** [comment]
Have you tried to run it only in electron? Does it worke in vanilla node.js on your machine?

**Comment by black-snow** [comment]
Got it running on my home machine (Win7 x64) with vanilla node.

**Comment by black-snow** [comment]
Sadly, with electron, it doesn't work on my home machine either. Got pretty much the same error.
I `cd`'ed into `app`, ran `npm install --save mdns` which ran just fine. I then used the exact same code snippet as in my vanilla version (the example from the project's README) and got prompted with this error.

![error](https://cloud.githubusercontent.com/assets/579733/12010541/da509ede-acaa-11e5-9558-8fe740ce692a.png)

**Comment by szwacz** [comment]
But you can't get electron-rebuild to work with this boilerplate?
The problem might be with two package.json files. Will do some testing.

**Comment by szwacz** [comment]
This way looks like it's working for me (for other module). Could you give it a try?

In file `electron-boilerplate/app/package.json` add lines:
```json
"scripts": {
  "postinstall": "electron-rebuild"
}
```

In command line:
```
cd electron-boilerplate
npm install electron-rebuild --save-dev
```

And then:
```
cd electron-boilerplate/app
npm install mdns --save
```

**Comment by black-snow** [comment]
Have you also added the example code for `mdns`? I've put it in background.js' onready:

```javascript
import mdns from 'mdns';

...

    // advertise a http server on port 4321
	var ad = mdns.createAdvertisement(mdns.tcp('http'), 4321);
	ad.start();

	// watch all http servers
	var browser = mdns.createBrowser(mdns.tcp('http'));
	browser.on('serviceUp', function(service) {
	  console.log("service up: ", service);
	});
	browser.on('serviceDown', function(service) {
	  console.log("service down: ", service);
	});
	browser.start();

	// discover all available service types
	var all_the_types = mdns.browseThemAll(); // all_the_types is just another browser...
```

**Comment by szwacz** [comment]
I'm a moron. The code have had put here earlier didn't work, I just assumed than no errors in console means it works.

`electron-rebuild` clearly doesn't like the fact that this project uses two `package.json`. But we can bypass this with JS API:

I've created file `electron-boilerplate/tasks/rebuild.js`:
```js
'use strict';

var childProcess = require('child_process');
var path = require('path');
var Q = require('q');
var electron = require('electron-prebuilt');
var rebuild = require('electron-rebuild');

var pathToElectronNativeModules = path.join(__dirname, '../app/node_modules');

var getElectronVersion = function () {
    var promisedExec = Q.denodeify(childProcess.exec);

    return promisedExec(electron + ' --version')
    .then(function (stdout) {
        var filterVersionOnly = /\d+\.\d+\.\d+/;
        return filterVersionOnly.exec(stdout)[0];
    });
};

rebuild.shouldRebuildNativeModules(electron)
.then(function (shouldBuild) {
    if (!shouldBuild) {
        return true;
    }

    console.log('Rebuilding native modules for Electron...');

    return getElectronVersion().then(function (version) {
        return rebuild.installNodeHeaders(version)
        .then(function () {
            return rebuild.rebuildNativeModules(version, pathToElectronNativeModules);
        });
    });
})
.then(function () {
    console.log('Rebuilding complete.');
})
.catch(function (err) {
    console.error("Rebuilding error!");
    console.error(err);
});
```

And in file `electron-boilerplate/app/package.json` added:
```json
"scripts": {
  "postinstall": "node ../tasks/rebuild"
}
```

The snippet about mdns you pasted here gives me error `Uncaught Exception: Error: dns service error: no such record`, what I'm not sure should happen, but this error looks like the library has started successfuly so I'm stopping here and asking you to try this out.

**Comment by black-snow** [comment]
I appreciate your help very much, thanks a lot!

Sadly, the rebuild task fails on my system. 

```javascript
return promisedExec(electron + ' --version')
    .then(function (stdout) {
```

`stdout` is `[ '', '' ]` on my system and thus `filterVersionOnly.exec(stdout)` is `null`. Executing `path\to\electron --version` manually returns `v0.35.4`, though. I hardcoded `getElectronVersion()` to return `0.35.4` and ... it worked! This is great!

I dunno why I get that weird value for `stdout` ... could we, alternatively, extract the prebuilt-version from the top-level *package.json* and use that?

How do we integrate your solution into the boilerplate? Running a rebuild after each `npm install` surely isn't necessary. We only need this for somewhat *native modules*, right?

**Comment by szwacz** [comment]
Hmm, that's odd. So `stdout` in your case is an array? Could you try if `childPrecess.spawn` instead of `childPrecess.exec` returns the version correctly?

**Comment by black-snow** [comment]
How do I do that? I'm not firm with *Q*. Changing `var promisedExec = Q.denodeify(childProcess.exec);` to `var promisedExec = Q.denodeify(childProcess.spawn);` leaves me with `[TypeError: Incorrect value of args option]`. I fiddled around but couldn't replace exec with spawn (though signatures seem the same).

**Comment by black-snow** [comment]
This: https://github.com/electronjs/electron-rebuild/issues/37#issuecomment-168773042 fixed the output for me.

rebuild.js:14 now looks like this:
```javascript
return promisedExec('SET ELECTRON_NO_ATTACH_CONSOLE=1 && ' + electron + ' --version')
```

**Comment by szwacz** [comment]
Native modules support landed in the boilerplate: https://github.com/szwacz/electron-boilerplate/blob/master/tasks/rebuild_native.js

I'm not happy with the fact that `postinstall` script isn't trigger when you do `npm install my_native_module --save` and you need to do extra step by typing `npm install` again to trigger it. If anyone has a better solution I'm all ears.

**Comment by black-snow** [comment]
I'm quite in a hurry right now but couldn't we just add another task, let's say `require_native` that does all that for us?
`npm run require_native arg` => `npm install --save arg && node run tasks/rebuild`

Thanks for your efforts!

**Comment by szwacz** [comment]
But this task will be run only once and then unnecessary. Or I'm missing something?

**Comment by black-snow** [comment]
yup, would be run whenever one needs to install a *native* module - it'd be just a shortcut
Why isn't `postinstall` triggered, btw?

**Comment by szwacz** [comment]
@black-snow You're right making it fully automatic wasn't that hard. It's done on `native-modules-installation` branch.

Commit: https://github.com/szwacz/electron-boilerplate/commit/be4300fea0612377ff2276e3071101cdf9fcda2f

Please give it a shot so I can make sure I didn't screw anything.

**Comment by black-snow** [comment]
Will take a look soon (& report back), thanks

**Comment by black-snow** [comment]
`npm run install-native mdns` gives me this:

```
{ [Error: spawn npm ENOENT]
  code: 'ENOENT',
  errno: 'ENOENT',
  syscall: 'spawn npm',
  path: 'npm',
  spawnargs: [ 'install', '--save', 'mdns' ] }
```

Might be related to https://github.com/nodejs/node-v0.x-archive/issues/2318

**Comment by szwacz** [comment]
Indeed. Fixed version should work.
https://github.com/szwacz/electron-boilerplate/commit/63a23626c72846015b4eb3d4167a18b726511e95

**Comment by black-snow** [comment]
@szwacz good work! Seems fine. Thanks
